test/functional/test_framework/v2_p2p.py

```diff
@@ -18,3 +100,147 @@ def v2_ecdh(priv, ellswift_theirs, ellswift_ours, initiating):
+
+    def respond_v2_handshake(self, response):
+        """Responder begins the v2 handshake by sending its ellswift bytes and garbage. However, the responder
+        sends this after having received at least one byte that mismatches 16-byte V1_PREFIX."""
+        while len(self.received_prefix) < 16:
+            byte = response.read(1)
+            # return b"" if we need to receive more bytes
+            if not byte:
+                return b""
+            self.received_prefix += byte
+            if self.received_prefix[-1] != V1_PREFIX[len(self.received_prefix) - 1]:
+                self.privkey_ours, self.ellswift_ours = ellswift_create()
+                garbage_len = random.randrange(4096)
+                self.sent_garbage = os.urandom(garbage_len)
+                logger.debug("sending %d bytes of garbage data" % garbage_len)
+                return self.ellswift_ours + self.sent_garbage
+        # return -1 to decide v1 only after all 16 bytes processed
+        return -1
+
+    def complete_handshake(self, response):
+        """ Instantiates the encrypted transport and
+        sends garbage terminator + optional decoy packets + transport version packet.
+        Done by both initiator and responder."""
+        received_prefix = b'' if self.initiating else self.received_prefix
+        ellswift_theirs = received_prefix + response.read(64 - len(received_prefix))
+        ecdh_secret = self.v2_ecdh(self.privkey_ours, ellswift_theirs, self.ellswift_ours, self.initiating)
+        self.initialize_v2_transport(ecdh_secret)
+        # Send garbage terminator
+        msg_to_send = self.peer['send_garbage_terminator']
+        # Optionally send decoy packets after garbage terminator.
+        aad = self.sent_garbage
+        for decoy_content_len in [random.randint(1, 100) for _ in range(10)]:
+            msg_to_send += self.v2_enc_packet(decoy_content_len * b'\x00', aad=aad, ignore=True)
+            aad = b''
+        # Send version packet.
+        msg_to_send += self.v2_enc_packet(TRANSPORT_VERSION, aad=aad)
+        return msg_to_send
+
+    def authenticate_handshake(self, response):
+        """ Ensures that the received optional decoy packets and transport version packet are authenticated.
+        Marks the v2 handshake as complete. Done by both initiator and responder.
+        Returns:
+            1. length of bytes that were processed so that recvbuf can be updated
+            2. True if the authentication was successful/more bytes need to be received and False otherwise
+        """
+        received_garbage = response[:16]
+        response = response[16:]
+        processed_length = len(received_garbage)
+        for i in range(4096):
+            if received_garbage[-16:] == self.peer['recv_garbage_terminator']:
+                # Receive, decode, and ignore version packet.
+                # This includes skipping decoys and authenticating the received garbage.
+                aad = received_garbage[:-16]
```

Comment:
nit: as `received_garbage` is ensured to have a maximum size of 16 bytes a few lines above, the slicing isn't needed here.
```suggestion
            if received_garbage == self.peer['recv_garbage_terminator']:
                # Receive, decode, and ignore version packet.
                # This includes skipping decoys and authenticating the received garbage.
                aad = received_garbage
```

Reply:
but there's also an else condition below where if the `received_garbage` doesn't match the garbage terminator, we keep appending 1 byte to `received_garbage` until there's a match with garbage terminator/we've exceed max garbage limit. so even though initially `received_garbage` length is 16 bytes, it can take values from 17 bytes, ..., 17 + 4095 bytes.

Reply:
Oh right, I missed the else-branch and assumed `received_garbage` wouldn't change in the loop (which wouldn't make much sense), nevermind.